Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,235 @@
+    ///     - systemMetricsType: The type of system metrics to use. If none is provided this defaults to
+    ///                          `LinuxSystemMetrics` on Linux platforms and `NOOPSystemMetrics` on all other platforms.
+    ///     - labels: The labels to use for generated system metrics.
+    public init(pollInterval interval: DispatchTimeInterval = .seconds(2), metricsType: SystemMetrics.Type? = nil, labels: SystemMetricsLabels) {
+        self.interval = interval
+        self.metricsType = metricsType
+        self.labels = labels
+    }
+}
+
+internal class SystemMetricsHandler {
+    fileprivate let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsHandler", qos: .background)
+    fileprivate let timeInterval: DispatchTimeInterval
+    fileprivate let systemMetricsType: SystemMetrics.Type
+    fileprivate let labels: SystemMetricsLabels
+    fileprivate let processId: Int
+    fileprivate var task: DispatchWorkItem?
+
+    init(options: SystemMetricsOptions) {
+        self.timeInterval = options.interval
+        if let systemMetricsType = options.metricsType {
+            self.systemMetricsType = systemMetricsType
+        } else {
+            #if os(Linux)
+            self.systemMetricsType = LinuxSystemMetrics.self
+            #else
+            self.systemMetricsType = NOOPSystemMetrics.self
+            #endif
+        }
+        self.labels = options.labels
+        #if os(Windows)
+        self.processId = 0
+        #else
+        self.processId = Int(getpid())
+        #endif
+
+        self.task = DispatchWorkItem(qos: .background, block: {
+            let metrics = self.systemMetricsType.init(pid: self.processId)
+            if let vmem = metrics.virtualMemoryBytes { Gauge(label: self.labels.label(for: \.virtualMemoryBytes)).record(vmem) }
+            if let rss = metrics.residentMemoryBytes { Gauge(label: self.labels.label(for: \.residentMemoryBytes)).record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: self.labels.label(for: \.maxFds)).record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: self.labels.label(for: \.openFds)).record(openFds) }
+        })
+
+        self.updateSystemMetrics()
+    }
+
+    internal func cancelSystemMetrics() {
+        self.task?.cancel()
+        self.task = nil
+    }
+
+    internal func updateSystemMetrics() {
+        self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+            guard let task = self.task else { return }
+            task.perform()
+            self.updateSystemMetrics()
+        }
+    }
+}
+
+public struct SystemMetricsLabels {
+    let prefix: String
+    let virtualMemoryBytes: String
```

Comment:
can we add a link to the spec that specified that those are the "system metrics" and also the right names? `Fds` sounds overly abbreviated and also UNIX-centric?